File: test-files/kotlin/bad_example.kt
// Hard-to-detect Kotlin issues that static analyzers miss without strict rules

fun process(input: String?) {
    // 1. Unsafe call on nullable
    println(input!!.length) // will throw NPE if input is null, not flagged unless Detekt's "NullCheck" rules are on

    // 2. Unused variable (hidden dead code)
    val temp = calculateSomething() // result never used, needs 'unusedVariable' rule
    // no warning unless linting for unused values

    // 3. Shadowed variable
    var total = 10
    if (true) {
        val total = 5 // shadows outer variable silently
        println(total) // prints 5, but outer 'total' stays 10
    }

    // 4. Ignored return value
    listOf(1, 2, 3).map { it * 2 } // return ignored — useless call unless rule "ignoredReturnValue" is on

    // 5. Silent integer division pitfall
    val ratio = 3 / 2 // = 1, not 1.5 — both are Int
    println("Ratio: $ratio")

    // 6. Data class equality confusion
    data class User(val name: String)
    val a = User("Nik")
    val b = User("Nik")
    if (a === b) println("Same!") // reference equality, not structural; won't trigger static warning

    // 7. Suspicious empty when branch
    val x = 5
    when (x) {
        1 -> println("One")
        2 -> println("Two")
        else -> {} // does nothing silently
    }

    // 8. Non-exhaustive when on enum
    enum class Mode { ON, OFF }
    val mode = Mode.ON
    when (mode) {
        Mode.ON -> println("On!") // missing OFF — compiles because no 'else'
    }

    // 9. Mutating a var parameter (bad style, hidden logic bug)
    var counter = 0
    fun add() { counter++ }
    add(); add(); add()
    println("Counter = $counter")

    // 10. Using == on floating point values
    val f1 = 0.1 + 0.2
    val f2 = 0.3
    if (f1 == f2) println("Equal!") // false due to precision, but no warning without 'FloatingPointEquality' rule
}

fun calculateSomething(): Int {
    return (1..10).sum()
}

fun main() {
    process(null)
}

File: test-files/swift/bad_example.swift
// Hard-to-detect Swift issues — compiles fine, but logically broken or misleading

import Foundation

func processData(_ input: String?) {
    // 1. Force unwrapping optional (runtime crash if nil)
    print("Length:", input!.count) // passes unless 'force_unwrapping' lint is on

    // 2. Unused result from map
    [1, 2, 3].map { $0 * 2 } // return value ignored — no warning by default

    // 3. Implicitly unwrapped optional misuse
    var name: String! = "Nikhil"
    name = nil
    print(name.count) // crash, but analyzer won't warn unless strict mode


    // 4. Shadowing variable
    let data = "Hello"
    if true {
        let data = "World" // shadows silently
        print(data)
    }
    print(data) // prints original, subtle logic confusion

    // 5. Equality on floating-point values
    let x = 0.1 + 0.2
    if x == 0.3 {
        print("Equal!") // false due to precision, not flagged unless 'floating_point_equality' rule is on
    }

    // 6. Unowned reference capture — potential dangling pointer
    class Foo {
        var closure: (() -> Void)?
        func setClosure() {
            closure = { [unowned self] in
                print("Foo instance:", self)
            }
        }
    }

    do {
        let foo = Foo()
        foo.setClosure()
        // foo goes out of scope -> closure now references deallocated object
    }

    // 7. Wrong collection API
    let dict = ["a": 1, "b": 2]
    for k in dict { // iterates tuple, not key
        print(k.0) // subtle — needs advanced rule 'for_in_over_dictionary'
    }

    // 8. Mutable default argument trap
    func appendItem(_ item: Int, to list: inout [Int] = [1, 2, 3]) {
        list.append(item) // modifies shared default across calls
        print(list)
    }
    var myList = [1, 2]
    appendItem(4, to: &myList)
    appendItem(5) // modifies default argument unexpectedly

    // 9. String indexing bug
    let text = "Swift🦅"
    let idx = text.index(text.startIndex, offsetBy: 5) // off-by-one due to Unicode graphemes
    print(text[idx]) // runtime error, but static analyzer sees nothing

    // 10. Capturing self strongly in closure
    class Bar {
        var value = 0
        func start() {
            Timer.scheduledTimer(withTimeInterval: 1.0, repeats: true) { _ in
                self.value += 1 // retain cycle, leaks Bar
            }
        }
    }
    Bar().start()
}

processData(nil)
